azurelinuxagent/ga/update.py

```diff
@@ -0,0 +1,278 @@
+# Windows Azure Linux Agent
+#
+# Copyright 2014 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.4+ and Openssl 1.0+
+#
+import os
+import json
+import time
+import subprocess
+import signal
+import sys
+import zipfile
+from azurelinuxagent.common.exception import UpdateError, ProtocolError
+import azurelinuxagent.common.logger as logger
+import azurelinuxagent.common.conf as conf
+from azurelinuxagent.common.event import add_event
+from azurelinuxagent.common.future import ustr
+import azurelinuxagent.common.utils.fileutil as fileutil
+import azurelinuxagent.common.utils.restutil as restutil
+from azurelinuxagent.common.version import AGENT_VERSION
+from azurelinuxagent.common.utils.textutil import Version
+from azurelinuxagent.common.osutil import get_osutil
+from azurelinuxagent.common.protocol import get_protocol_util
+
+def get_update_handler():
+    return UpdateHandler()
+
+GA_DIR = 'GuestAgent' # Dir name for guest agent bin and data
+GA_ERR = 'error' # File name for guest agent error record
+MAX_FAILURE = 3 # Max failure allowed for guest agent before blacklisted
+RETAIN_INTERVAL = 24 * 60 * 60 # Retain interval for black list
+
+"""
+Handles self update logic
+"""
+class UpdateHandler(object):
+
+    def __init__(self):
+        self.osutil = get_osutil()
+        self.protocol_util = get_protocol_util()
+        self.last_etag = None
+        self.agents = []
+        self.error_record = {}
+
+    def run(self):
```

Comment:
@ahmetalpbalkan You are correct. The current code integrates update as a pre-step to extension handling. We could add flags to tweak that.